crates/node/src/vmm/qemu.rs

```diff
@@ -309,19 +303,54 @@ impl Provider for Qemu {
                     std::process::exit(1);
                 }
 
-                match Qmp::new(format!("localhost:{qmp_port}")).await {
-                    Ok(c) => client = Some(c),
-                    Err(err) => {
-                        retry_count += 1;
-                        sleep(Duration::from_millis(10)).await;
+                match timeout(
+                    QMP_CONNECT_TIMEOUT,
+                    Qmp::new(format!("localhost:{qmp_port}")),
+                )
+                .await
+                {
+                    Ok(connect) => match connect {
+                        Ok(clnt) => client = Some(clnt),
+                        Err(err) => {
+                            // Connection was refused. QEMU not started yet.
+                            retry_count += 1;
+                            sleep(Duration::from_millis(10)).await;
+                        }
+                    },
+                    Err(_) => {
+                        if retry_count < 100 {
+                            tracing::warn!(
+                                "tx: {} - QEMU QMP connect timeout; retrying once",
+                                tx_hash
+                            );
+                            retry_count = 99;
+                        } else {
+                            tracing::error!(
+                                "tx: {} - QEMU QMP connect timeout. Terminating VM.",
+                                tx_hash
+                            );
+                            let cid = qemu_vm_handle.cid;
+                            qemu_vm_handle
+                                .child
+                                .as_mut()
+                                .ok_or(std::io::Error::other(
+                                    "No child process defined for this handle",
+                                ))
+                                .and_then(|p| {
+                                    p.kill()?;
+                                    p.wait()
+                                })?;
+
+                            self.vm_registry.remove(&cid);
+                            self.cid_allocations.remove(&cid);
+                            return Err(eyre!("Failed to connect to QEMU QMP"));
+                        }
                     }
                 };
             }
             client.unwrap()
         };
 
-        qmp_client.system_reset().await?;
-
```

Comment:
No need for extra VM reboot anymore, since we don't attach the disk during runtime via QMP. When I saw this, I got thinking that _maybe_ this could also have contributed to the issues with the gRPC-over-VSOCK :thinking: 

I bet that with right timing, the VM can probably _just_ start and connect, until it already gets smacked in the head for reboot. That would probably leave the server in very confused state.